### DISCUSSION TITLE
A way to for-see the actions available in a state machine - using x-state for workflow management, like JIRA workflow

### DISCUSSION THREAD

**Discussion by pranav-aggarwal** [statelyai/xstate#1618] [Q&A]
I am using x-state as workflow and action guards to for-see, next available actions for the users in a given state.

Say, i am in a workflow to publish a response to a question, and before showcasing the publish action to the end user, i need to apply a rule, to check if response to question already being made or not, if response already exist, do not allow action on that state.

For this rule check, i have to make persistence queries, but my guard is coming out of async mode. I am not able to use invoke functionality here to for-see actions.

```{"new": {
                "id": "new",
                "on": {
                    "DRAFT": {
                        "target": "draft",
                        "cond": {
                            "type": "accessGrant",
                            "accessGroup": [
                                "contributorRole",
                                "notanonymous"
                            ],
                            "rules":  {
                                "all": [
                                    {
                                        "fact": "assignmentAlreadySubmitted",
                                        "value": false,
                                        "operator": "equal"
                                    }
                                ]
                            }
                        }
                    },
                    "SUBMIT": {
                        "target": "submitted",
                        "cond": {
                            "type": "accessGrant",
                            "accessGroup": [
                                "contributorRole",
                                "notreviewerRole"
                            ],
                            "rules":  {
                                "all": [
                                    {
                                        "fact": "assignmentAlreadySubmitted",
                                        "value": false,
                                        "operator": "equal"
                                    }
                                ]
                            }
                        }
                    }
                },
                "meta": {
                    "actions": [
                        {
                            "name": "DRAFT",
                            "title": "Save As Draft",
                            "order": 0,
                            "editable": true,
                            "primary": false,
                            "changeViewOnSuccess": false,
                            "successMessage": "Draft saved.",
                            "failureMessage": "Unable to save draft."
                        },
                        {
                            "name": "SUBMIT",
                            "title": "Submit",
                            "order": 0,
                            "primary": true,
                            "editable": false,
                            "changeViewOnSuccess": true,
                            "successMessage": "Request submitted, a moderator will look into it.",
                            "failureMessage": "Unable to submit."
                        }
                    ]
                }
            }}```

**Comment by pranav-aggarwal** [comment]
@davidkpiano  Do you think we can resolve this problem, other wise i have a workaround to pre-calculate the facts and add them in context before entering the state machine. Issue with this approach is with the time code will become unmanageable.

**Reply by pranav-aggarwal** [reply]
Adding it as my notes: 

https://workflowengine.io/blog/workflow-engine-vs-state-machine/ 

It has defined a very basic architectural difference between FSM and BPMN2  models. But i think, why can't there be an heterogeneous solution. In my usage of XState i have to keep the state of a document, which has some specific actions defined in that state. What i was looking is to filter those actions based on the conditions defined in the actions. 

What restricting us to allow synchronous conditions if not actions, (this similar issue i faced during synchronous actions and some how managed to hack around it using my custom logic on top of XState).
 
I am revisiting my decision, if XState was bad choice i made, as per the definition in above link?

**Reply by Andarist** [reply]
> I am revisiting my decision, if xstate was bad choice i made, as per the definition in above link?

The article assumes much more simpler state machine engines than XState. The main difference is that XState actually can model stepped workflows, parallel states and hierarchical ones and the article assumes that a state engine cant do those. It assumes simple state machines whereas XState implements statecharts that are far way more powerful.

As to your original question - could u prepare a codesandbox that would illustrate your problem? It would be easier for me to see whats the problem at hand and how you have been trying to approach it so far

**Comment by sudoStatus200** [comment]
@pranav-aggarwal  were you able to build something like jira workflow using XState ? I also need to build something similiar. Although I think XState is right tool for this but I want to be sure and probably try one of BPMN engine as well.
